src/sgdnet.cpp

```diff
@@ -329,86 +137,81 @@ Rcpp::List SetupSgdnet(T&                   x,
   // Maximum of sums of squares over samples
   double max_squared_sum = ColNormsMax(x);
 
-  std::vector<double> step_size = family->StepSize(max_squared_sum,
-                                                   alpha,
-                                                   fit_intercept,
-                                                   n_samples);
+  vector<double> step_size =
+    family->StepSize(max_squared_sum, alpha, fit_intercept, n_samples);
 
   // Check if we need the nontrivial prox
   // TODO(jolars): allow more proximal operators
-  sgdnet::ProxFactory prox_factory;
-  std::string prox_choice = "soft_threshold";
-  std::unique_ptr<sgdnet::Prox> prox = prox_factory.NewProx(prox_choice);
+  string prox_choice = "soft_threshold";
+
+  unique_ptr<sgdnet::Prox> prox;
+
+  if (prox_choice == "soft_threshold") {
+    prox = unique_ptr<sgdnet::Prox>(new sgdnet::SoftThreshold());
+  }
 
   // Setup intercept vector
-  std::vector<double> intercept(n_classes, 0.0);
-  std::vector< std::vector<double> > intercept_archive(n_penalties);
+  vector<double> intercept(n_classes);
+  vector<vector<double>> intercept_archive;
 
   // Setup weights matrix and weights archive
-  std::vector<double> weights(n_features*n_classes, 0.0);
-  std::vector< std::vector<double> > weights_archive;
+  vector<double> weights(n_features*n_classes);
+  vector<vector<double>> weights_archive;
 
   // Sum of gradients for weights
-  std::vector<double> sum_gradient(n_features*n_classes);
-
-  // Sum of gradients for intercept
-  std::vector<double> intercept_sum_gradient(n_classes, 0.0);
+  vector<double> sum_gradient(n_features*n_classes);
 
   // Gradient memory
-  std::vector<double> gradient_memory(n_samples*n_classes);
+  vector<double> gradient_memory(n_samples*n_classes);
 
-  // Keep track of the number of as well as which samples are seen
-  std::vector<bool> seen(n_samples, false);
-  std::size_t n_seen = 0;
+  // Sum of gradients for intercept
+  vector<double> sum_gradient_intercept(n_classes);
 
   // Keep keep track of successes for each penalty
-  std::vector<unsigned int> return_codes;
+  vector<unsigned> return_codes;
 
   // Setup a vector of loss vectors to return upon exit
-  std::vector< std::vector<double> > losses_archive;
-  std::vector<double> losses;
+  vector<vector<double>> losses_archive;
 
   // Keep track of number of iteratios per penalty
-  std::size_t n_iter = 0;
+  unsigned n_iter = 0;
 
   // Null deviance on scaled y for computing deviance ratio
   double null_deviance_scaled = family->NullDeviance(y);
-  std::vector<double> deviance_ratio;
+  vector<double> deviance_ratio;
   deviance_ratio.reserve(n_penalties);
 
-  // Store the nonzero indices of each sample
-  std::vector< std::vector<std::size_t> > nonzero_indices_storage(n_samples);
-
   // Fit the path of penalty values
-  for (std::size_t penalty_ind = 0; penalty_ind < n_penalties; ++penalty_ind) {
+  for (unsigned penalty_ind = 0; penalty_ind < n_penalties; ++penalty_ind) {
```

Comment:
Possibly think about changing this to something like `regularization_ind`. `penalty_ind` is definitely mathematically correct, but IME folks generally use "penalty" to refer to the base function and "lambda" to refer to the coefficient. E.g., in `argmin_{beta} |y - Xbeta|_2^2 + 3 |beta|_1` people would say it's L1-penalty, not a 3*L1 penalty.  (It's a distinction without a difference mathematically, but it's what people do...)

Reply:
You're right. I'll change it to lambda_ind and n_lambda.